src/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteSslContext.cs

```diff
@@ -0,0 +1,345 @@
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (null != _sslContext)
+                {
+                    _sslContext.Dispose();
+                    _sslContext = null;
+                }
+
+                _toConnection = null;
+                _fromConnection = null;
+                _writeCallback = null;
+                _readCallback = null;
+            }
+
+            base.Dispose(disposing);
+        }
+
+        private unsafe int WriteToConnection(void* connection, byte* data, void** dataLength)
+        {
+            ulong toWrite = (ulong)*dataLength;
+            byte* readFrom = data;
+
+            lock (_toConnection)
+            {
+                while (toWrite > 0)
+                {
+                    _toConnection.Enqueue(*readFrom);
+                    readFrom++;
+                    toWrite--;
+                }
+            }
+
+            // Since we can enqueue everything, no need to re-assign *dataLength.
+            const int noErr = 0;
+            return noErr;
+        }
+
+        private unsafe int ReadFromConnection(void* connection, byte* data, void** dataLength)
+        {
+            const int noErr = 0;
+            const int errSSLWouldBlock = -9803;
+
+            ulong toRead = (ulong)*dataLength;
+
+            if (toRead == 0)
+            {
+
+                return noErr;
+            }
+
+            uint transferred = 0;
+
+            lock (_fromConnection)
+            {
+
+                if (_fromConnection.Count == 0)
+                {
+
+                    *dataLength = (void*)0;
+                    return errSSLWouldBlock;
+                }
+
+                byte* writePos = data;
+
+                while (transferred < toRead && _fromConnection.Count > 0)
+                {
+                    *writePos = _fromConnection.Dequeue();
+                    writePos++;
+                    transferred++;
+                }
+            }
+
+            *dataLength = (void*)transferred;
+            return noErr;
+        }
+
+        internal void Write(byte[] buf, int offset, int count)
+        {
+            Debug.Assert(buf != null);
+            Debug.Assert(offset >= 0);
+            Debug.Assert(count >= 0);
+            Debug.Assert(count <= buf.Length - offset);
+
+
+            lock (_fromConnection)
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    _fromConnection.Enqueue(buf[offset + i]);
+                }
+            }
+
+        }
+
+        internal int BytesReadyForConnection => _toConnection.Count;
+
+        internal byte[] ReadPendingWrites()
+        {
+            lock (_toConnection)
+            {
+                if (_toConnection.Count == 0)
+                {
+                    return null;
+                }
+
+                byte[] data = _toConnection.ToArray();
```

Comment:
Maybe I don't understand the design here... but there seems to be a bunch of things here that could be very expensive.